File: packages/react-native-fipa/android/src/main/java/com/deviceattestation/DeviceAttestationFirstPartyTransport.kt
package com.deviceattestation

import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReactApplicationContext

/**
 * Foreground protocol transport. Android enrollment must use the challenge-bound key module; the
 * iOS prepareDpop entry point can never mint an unattested key.
 */
internal class DeviceAttestationFirstPartyTransport(context: ReactApplicationContext) :
    NativeFirstPartyTransportSpec(context) {
    private val client = FirstPartyHTTPClient()
    private val keys = AndroidAttestedKey(context.applicationContext)
    private val operations =
        BridgeOperations(
            1,
            64,
            { error -> (error as? NativeFailure)?.code ?: "http_unavailable" },
            "The native operation could not be completed.",
        )

    override fun randomToken(promise: Promise) =
        operations.execute(promise) { FirstPartyCrypto.randomToken() }

    override fun transaction(promise: Promise) =
        operations.execute(promise) {
            val verifier = FirstPartyCrypto.randomToken()
            Arguments.createMap().apply {
                putString("id", FirstPartyCrypto.randomToken())
                putString("verifier", verifier)
                putString("challenge", FirstPartyCrypto.sha256Utf8(verifier))
            }
        }

    override fun prepareDpop(alias: String, promise: Promise) =
        operations.execute(promise) { throw NativeFailure("key_enrollment_required") }

    override fun inspectDpop(alias: String, promise: Promise) =
        operations.execute(promise) { keys.inspect(alias) ?: throw NativeFailure("key_missing") }

    override fun removeDpop(alias: String, expectedThumbprint: String, promise: Promise) =
        operations.execute(promise) {
            keys.remove(alias, expectedThumbprint)
            null
        }

    override fun signDpop(
        alias: String,
        expectedThumbprint: String,
        url: String,
        method: String,
        accessToken: String?,
        nonce: String?,
        promise: Promise,
    ) =
        operations.execute(promise) {
            keys.sign(alias, expectedThumbprint, url, method, accessToken, nonce)
        }

    override fun send(
        requestId: String,
        url: String,
        method: String,
        headersJSON: String,
        body: String?,
        maximumResponseBytes: Double,
        timeoutMilliseconds: Double,
        allowInsecureLoopback: Boolean,
        promise: Promise,
    ) {
        if (!operations.begin(promise)) return
        client
            .send(
                requestId,
                url,
                method,
                headersJSON,
                body,
                maximumResponseBytes,
                timeoutMilliseconds,
                allowInsecureLoopback,
            )
            .whenComplete { response, error ->
                if (error != null) operations.reject(promise, error)
                else
                    operations.resolve(
                        promise,
                        Arguments.createMap().apply {
                            putString("url", response.url)
                            putInt("status", response.status)
                            putString("headersJSON", response.headersJSON)
                            putString("body", response.body)
                        },
                    )
            }
    }

    override fun cancel(requestId: String, promise: Promise) {
        client.cancel(requestId)
        promise.resolve(null)
    }

    override fun openBrowser(
        requestId: String,
        url: String,
        redirectUri: String,
        timeoutMilliseconds: Double,
        allowInsecureLoopback: Boolean,
        promise: Promise,
    ) {
        if (!operations.begin(promise)) return
        AndroidFirstPartyBrowser.open(
            this,
            reactApplicationContext,
            BrowserRequest(requestId, url, redirectUri, timeoutMilliseconds, allowInsecureLoopback),
            { operations.invalidated },
        ) { callback, error ->
            if (error != null) operations.reject(promise, error)
            else operations.resolve(promise, callback)
        }
    }

    override fun cancelBrowser(requestId: String, promise: Promise) {
        AndroidFirstPartyBrowser.cancel(this, requestId) { promise.resolve(null) }
    }

    override fun invalidate() {
        operations.invalidate {
            AndroidFirstPartyBrowser.invalidate(this)
            client.close()
        }
        super.invalidate()
    }
}

File: packages/react-native-fipa/ios/DeviceAttestationFirstPartyTransport.swift
import Foundation
import React

@objc(DeviceAttestationFirstPartyTransport)
class DeviceAttestationFirstPartyTransport: NSObject {
  private static let client = FirstPartyHTTPClient()
  @objc static func requiresMainQueueSetup() -> Bool { false }

  @objc func randomToken(_ resolve: RCTPromiseResolveBlock, rejecter _: RCTPromiseRejectBlock) {
    resolve(FirstPartyCrypto.randomToken())
  }
  @objc func transaction(_ resolve: RCTPromiseResolveBlock, rejecter _: RCTPromiseRejectBlock) {
    resolve(FirstPartyCrypto.transaction())
  }
  @objc func prepareDpop(_ alias: String, resolver resolve: RCTPromiseResolveBlock, rejecter reject: RCTPromiseRejectBlock) {
    keyCall(resolve, reject) { try FirstPartyDpopKey.prepare(alias: alias) }
  }
  @objc func inspectDpop(_ alias: String, resolver resolve: RCTPromiseResolveBlock, rejecter reject: RCTPromiseRejectBlock) {
    keyCall(resolve, reject) { try FirstPartyDpopKey.inspect(alias: alias) }
  }
  @objc func removeDpop(_ alias: String, expectedThumbprint: String, resolver resolve: RCTPromiseResolveBlock, rejecter reject: RCTPromiseRejectBlock) {
    keyCall(resolve, reject) { try FirstPartyDpopKey.remove(alias: alias, expectedThumbprint: expectedThumbprint); return nil }
  }
  @objc func signDpop(_ alias: String, expectedThumbprint: String, url: String, method: String,
                     accessToken: String?, nonce: String?, resolver resolve: RCTPromiseResolveBlock, rejecter reject: RCTPromiseRejectBlock) {
    keyCall(resolve, reject) { try FirstPartyDpopKey.proof(alias: alias, expectedThumbprint: expectedThumbprint,
      url: url, method: method, accessToken: accessToken, nonce: nonce) }
  }
  private func keyCall(_ resolve: RCTPromiseResolveBlock, _ reject: RCTPromiseRejectBlock, _ operation: () throws -> Any?) {
    do { resolve(try operation()) }
    catch let error as FirstPartyKeyError { reject(error.rawValue, "Unable to use the device signing key.", nil) }
    catch { reject(FirstPartyKeyError.failed.rawValue, "Unable to use the device signing key.", nil) }
  }
  @objc func cancel(_ requestId: String, resolver resolve: RCTPromiseResolveBlock, rejecter _: RCTPromiseRejectBlock) {
    Self.client.cancel(id: requestId); resolve(nil)
  }
  @objc func openBrowser(_ requestId: String, url: String, redirectUri: String, timeoutMilliseconds: Double,
                        allowInsecureLoopback: Bool, resolver resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
    DispatchQueue.main.async {
      FirstPartyBrowser.shared.open(id: requestId, url: url, redirectUri: redirectUri,
        timeoutMilliseconds: timeoutMilliseconds, allowInsecureLoopback: allowInsecureLoopback) { result in
        switch result {
        case .success(let callback): resolve(callback)
        case .failure(let error): reject(error.rawValue, "Unable to complete browser authentication.", nil)
        }
      }
    }
  }
  @objc func cancelBrowser(_ requestId: String, resolver resolve: @escaping RCTPromiseResolveBlock, rejecter _: RCTPromiseRejectBlock) {
    DispatchQueue.main.async { FirstPartyBrowser.shared.cancel(id: requestId); resolve(nil) }
  }
  @objc func send(_ requestId: String, url: String, method: String, headersJSON: String, body: String?,
                  maximumResponseBytes: Double, timeoutMilliseconds: Double, allowInsecureLoopback: Bool,
                  resolver resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
    guard headersJSON.utf8.count <= 32768,
          let headers = try? JSONDecoder().decode([String: String].self, from: Data(headersJSON.utf8)),
          maximumResponseBytes.isFinite, maximumResponseBytes > 0, maximumResponseBytes <= 1_048_576,
          maximumResponseBytes.rounded(.towardZero) == maximumResponseBytes,
          timeoutMilliseconds.isFinite, timeoutMilliseconds >= 1000, timeoutMilliseconds <= 120000,
          timeoutMilliseconds.rounded(.towardZero) == timeoutMilliseconds else {
      reject(FirstPartyHTTPError.invalidRequest.rawValue, "Unable to complete authentication request.", nil); return
    }
    Self.client.send(id: requestId, url: url, method: method, headers: headers, body: body,
      maximumResponseBytes: Int(maximumResponseBytes), timeoutMilliseconds: Int(timeoutMilliseconds),
      allowInsecureLoopback: allowInsecureLoopback) { result in
        switch result {
        case .success(let response):
          guard let data = try? JSONEncoder().encode(response.headers), let headersJSON = String(data: data, encoding: .utf8) else {
            reject(FirstPartyHTTPError.invalidResponse.rawValue, "Unable to complete authentication request.", nil); return
          }
          resolve(["url": response.url, "status": response.status, "headersJSON": headersJSON, "body": response.body])
        case .failure(let error):
          reject(error.rawValue, "Unable to complete authentication request.", nil)
        }
      }
  }
}
